src/theory/trust_substitutions.h

```diff
@@ -0,0 +1,67 @@
+/*********************                                                        */
+/*! \file trust_substitutions.h
+ ** \verbatim
+ ** Top contributors (to current version):
+ **   Andrew Reynolds
+ ** This file is part of the CVC4 project.
+ ** Copyright (c) 2009-2020 by the authors listed in the file AUTHORS
+ ** in the top-level source directory and their institutional affiliations.
+ ** All rights reserved.  See the file COPYING in the top-level source
+ ** directory for licensing information.\endverbatim
+ **
+ ** \brief Trust substitutions
+ **/
+
+#include "cvc4_private.h"
+
+#ifndef CVC4__THEORY__TRUST_SUBSTITUTIONS_H
+#define CVC4__THEORY__TRUST_SUBSTITUTIONS_H
+
+#include "context/context.h"
+#include "expr/proof_node_manager.h"
+#include "theory/substitutions.h"
+#include "theory/trust_node.h"
+
+namespace CVC4 {
+namespace theory {
+
+/**
+ * A layer on top of SubstitutionMap that tracks proofs.
+ */
+class TrustSubstitutionMap
+{
+ public:
+  TrustSubstitutionMap(context::Context* c, ProofNodeManager* pnm);
+  /** Gets a reference to the underlying substitution map */
+  SubstitutionMap& get();
+  /**
+   * Add substitution x -> t, where pg can provide a closed proof of (= x t)
+   * in the remainder of this user context.
+   */
+  void addSubstitution(TNode x, TNode t, ProofGenerator* pg = nullptr);
+  /**
+   * Add substitution x -> t, which was derived from the proven field of
+   * trust node tn. This ensures that we add a substitution with a proof
+   * based on transforming the tn.getProven() to (= x t) if tn has a
+   * non-null proof generator; otherwise if tn has no proof generator
+   * we simply add the substitution.
+   */
+  void addSubstitutionSolved(TNode x, TNode t, TrustNode tn);
```

Comment:
I don't follow the name. Why not overload `addSubstitution`?

Reply:
This indicates that we had a literal `lit` inside of `tn`, and it was put into solved form `x = t`.

I think this is a bit of a different interface than the one above, which says i can prove `x = t`.  This method is with respect to a previous formula and puts it into solved form. I will add some more comments about this.

Reply:
I see. This "solved form" bit clarifies it. I guess it's not necessary to add comments on the calls.